CL/cl_half.h

```diff
@@ -0,0 +1,458 @@
+/*******************************************************************************
+ * Copyright (c) 2019 The Khronos Group Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and/or associated documentation files (the
+ * "Materials"), to deal in the Materials without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Materials, and to
+ * permit persons to whom the Materials are furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Materials.
+ *
+ * MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+ * KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+ * SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+ *    https://www.khronos.org/registry/
+ *
+ * THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+ ******************************************************************************/
+
+/**
+ * This is a header-only utility library that provides OpenCL host code with
+ * routines for converting to/from cl_half values.
+ *
+ * Example usage:
+ *
+ *    #include <CL/cl_half.h>
+ *    ...
+ *    cl_half h = cl_float_to_half(0.5f, CL_HALF_RTE);
+ *    cl_float f = cl_half_to_float(h);
+ */
+
+#ifndef __CL_HALF_H
+#define __CL_HALF_H
+
+#include <CL/cl_platform.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+/**
+ * Rounding mode used when converting to cl_half.
+ */
+typedef enum
+{
+  CL_HALF_RTE, // round to nearest even
+  CL_HALF_RTZ, // round towards zero
+  CL_HALF_RTP, // round towards positive infinity
+  CL_HALF_RTN, // round towards negative infinity
+} cl_half_rounding_mode;
+
+
+/* Private utility macros. */
+#define CL_HALF_EXP_MASK 0x7C00
+#define CL_HALF_MAX_FINITE_MAG 0x7BFF
+
+
+/*
+ * Utility to deal with values that overflow when converting to half precision.
+ */
+static inline cl_half cl_half_handle_overflow(cl_half_rounding_mode rounding_mode,
```

Comment:
Confirming: the functions in this file use an underscore-delimited "snake case" naming convention as opposed to the OpenCL API "camel case" convention.  Is that OK?  Similar to the comment above, perhaps we should consider a different prefix to differentiate from standard API functions, if nothing else?

Reply:
So the choice for snake case was in part to differentiate from the API functions. A prefix is a good idea though if we can find one we're happy with. As above, we could have `clxFloatToHalf`, `cluFloatToHalf` or `oclFloatToHalf`, but I'm very much open to suggestions (since none of these look nice to me!).

Reply:
Thinking about this a bit more, I think I'd prefer to use a different prefix first and foremost.  I don't have a strong opinion regarding snake case or camel case, since the different prefix alone will ensure no conflicts with existing or future APIs or typedefs or enums.

I'm fine with either `clu` / `CLU`, or `clx` / `CLX`, or a different prefix.  I suppose if I had to choose I'd go with `clu`, even given the old [Computing Language Utility](https://github.com/Computing-Language-Utility/CLU).

```c
    cl_float f = 1.0f;
    cl_half h = clu_float_to_half(f, CLU_HALF_RTE);
    f = clu_half_to_float(h);
```

We should consider a different prefix for the name of this file, too.